From 2d1be603655e450ac666e6003ae3534074c334f2 Mon Sep 17 00:00:00 2001
From: repo sync <hejiean@eswincomputing.com>
Date: Fri, 27 Jun 2025 10:22:09 +0800
Subject: [PATCH 586/652] fix:dmabuf dma_resv lock was not held

Changelogs:
1. memory replace dma_buf_vmap, dma_buf_vunmap,
dma_buf_map_attachment,dma_buf_unmap_attachment to _unlocked version
2. fix error handle lack unmap resource.
3. fix npu,hae,gpu dma_buf api calling without dma buf reserve lock locked

Signed-off-by: hejiean <hejiean@eswincomputing.com>
---
 .../server/env/linux/physmem_dmabuf.c         |  5 ++++
 drivers/memory/eswin/es_mmz_vb/mmz_vb.c       | 13 +++++-----
 .../dmabuf-heap-import-helper.c               | 13 ++++++----
 drivers/soc/eswin/ai_driver/npu/npu_main.c    | 24 ++++++++++++-------
 .../default/gc_hal_kernel_allocator_dmabuf.c  | 12 ++++++++++
 5 files changed, 48 insertions(+), 19 deletions(-)

diff --git a/drivers/gpu/drm/img/img-volcanic/services/server/env/linux/physmem_dmabuf.c b/drivers/gpu/drm/img/img-volcanic/services/server/env/linux/physmem_dmabuf.c
index 97f84d6095bc..e82b0d5cebe6 100644
--- a/drivers/gpu/drm/img/img-volcanic/services/server/env/linux/physmem_dmabuf.c
+++ b/drivers/gpu/drm/img/img-volcanic/services/server/env/linux/physmem_dmabuf.c
@@ -1382,7 +1382,10 @@ PMRAcquireKernelMappingDataDmaBuf(PMR_IMPL_PRIVDATA pvPriv,
 		goto fail;
 	}
 
+	_pvr_dma_resv_lock(psDmaBuf);
 	err = dma_buf_vmap(psDmaBuf, &psPrivData->sMap);
+	_pvr_dma_resv_unlock(psDmaBuf);
+
 	if (err != 0 || psPrivData->sMap.vaddr == NULL)
 	{
 		eError = PVRSRV_ERROR_PMR_NO_KERNEL_MAPPING;
@@ -1411,7 +1414,9 @@ static void PMRReleaseKernelMappingDataDmaBuf(PMR_IMPL_PRIVDATA pvPriv,
 	struct dma_buf *psDmaBuf = psPrivData->psAttachment->dmabuf;
 	int err;
 
+	_pvr_dma_resv_lock(psDmaBuf);
 	dma_buf_vunmap(psDmaBuf, &psPrivData->sMap);
+	_pvr_dma_resv_unlock(psDmaBuf);
 
 	do {
 		err = dma_buf_end_cpu_access(psDmaBuf, DMA_BIDIRECTIONAL);
diff --git a/drivers/memory/eswin/es_mmz_vb/mmz_vb.c b/drivers/memory/eswin/es_mmz_vb/mmz_vb.c
index 76f440b124d9..4e45d6bf4ffd 100644
--- a/drivers/memory/eswin/es_mmz_vb/mmz_vb.c
+++ b/drivers/memory/eswin/es_mmz_vb/mmz_vb.c
@@ -40,6 +40,7 @@
 #include <linux/delay.h>
 #include <linux/hashtable.h>
 #include <linux/es_proc.h>
+#include <linux/dma-resv.h>
 #include <linux/dmabuf-heap-import-helper.h>
 #include "include/linux/mmz_vb_type.h" /*must include before es_vb_user.h*/
 #include <uapi/linux/es_vb_user.h>
@@ -103,7 +104,7 @@ static int vb_blk_to_pool(struct esVB_BLOCK_TO_POOL_CMD_S *blkToPoolCmd);
 static int vb_get_blk_offset(struct esVB_GET_BLOCKOFFSET_CMD_S *getBlkOffsetCmd);
 static int vb_split_dmabuf(struct esVB_SPLIT_DMABUF_CMD_S *splitDmabufCmd);
 static int vb_get_dmabuf_refcnt(struct esVB_DMABUF_REFCOUNT_CMD_S *getDmabufRefCntCmd);
-static int vb_retrieve_mem_node(struct esVB_RETRIEVE_MEM_NODE_CMD_S *retrieveMemNodeCmd, 
+static int vb_retrieve_mem_node(struct esVB_RETRIEVE_MEM_NODE_CMD_S *retrieveMemNodeCmd,
 	eic770x_memory_type_t *p_mem_type);
 static int vb_get_dmabuf_size(struct esVB_DMABUF_SIZE_CMD_S *getDmabufSizeCmd);
 static int mmz_vb_pool_exit(void);
@@ -1473,7 +1474,7 @@ static int vb_ioctl_uninit_config(void __user *user_cmd)
 static int vb_ioctl_flush_all(void __user *user_cmd)
 {
 	EIC770X_LOGICAL_MEM_NODE_E nid;
-	
+
 	if (copy_from_user(&nid, user_cmd, sizeof(EIC770X_LOGICAL_MEM_NODE_E))) {
 		return -EFAULT;
 	}
@@ -2268,7 +2269,7 @@ static int vb_get_dmabuf_refcnt(struct esVB_DMABUF_REFCOUNT_CMD_S *getDmabufRefC
 	return ret;
 }
 
-static int do_vb_retrive_mem_node(struct dma_buf *dmabuf, 
+static int do_vb_retrive_mem_node(struct dma_buf *dmabuf,
 	EIC770X_LOGICAL_MEM_NODE_E *p_nid, eic770x_memory_type_t *p_mem_type)
 {
 	int ret = 0;
@@ -2286,7 +2287,7 @@ static int do_vb_retrive_mem_node(struct dma_buf *dmabuf,
 		return ret;
 	}
 
-	sgt = dma_buf_map_attachment(attach, DMA_BIDIRECTIONAL);
+	sgt = dma_buf_map_attachment_unlocked(attach, DMA_BIDIRECTIONAL);
 	if (IS_ERR(sgt)) {
 		ret = PTR_ERR(sgt);
 		dma_buf_detach(dmabuf, attach);
@@ -2298,7 +2299,7 @@ static int do_vb_retrive_mem_node(struct dma_buf *dmabuf,
 	pfn = page_to_pfn(page);
 	arch_get_mem_node_and_type(pfn, p_nid, p_mem_type);
 
-	dma_buf_unmap_attachment(attach, sgt, DMA_BIDIRECTIONAL);
+	dma_buf_unmap_attachment_unlocked(attach, sgt, DMA_BIDIRECTIONAL);
 	/* detach */
 	dma_buf_detach(dmabuf, attach);
 	/* put dmabuf back */
@@ -2308,7 +2309,7 @@ static int do_vb_retrive_mem_node(struct dma_buf *dmabuf,
 	return ret;
 }
 
-static int vb_retrieve_mem_node(struct esVB_RETRIEVE_MEM_NODE_CMD_S *retrieveMemNodeCmd, 
+static int vb_retrieve_mem_node(struct esVB_RETRIEVE_MEM_NODE_CMD_S *retrieveMemNodeCmd,
 	eic770x_memory_type_t *p_mem_type)
 {
 	int ret = 0;
diff --git a/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c b/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c
index f0d67b27ae4b..c4ce2e01a61f 100644
--- a/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c
+++ b/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c
@@ -28,6 +28,7 @@
 #include <linux/miscdevice.h>
 #include <linux/dma-map-ops.h>
 #include <linux/dma-heap.h>
+#include <linux/dma-resv.h>
 #include <linux/dmabuf-heap-import-helper.h>
 
 static struct device *split_dmabuf_dev;
@@ -214,7 +215,7 @@ static struct heap_mem *dmabuf_heap_import(struct heap_root *root, int fd)
         goto clean_up;
     }
 
-    sgt = dma_buf_map_attachment(attach, DMA_BIDIRECTIONAL);
+    sgt = dma_buf_map_attachment_unlocked(attach, DMA_BIDIRECTIONAL);
 	if (IS_ERR(sgt)) {
 		ret = PTR_ERR(sgt);
 		goto fail_detach;
@@ -248,6 +249,7 @@ static struct heap_mem *dmabuf_heap_import(struct heap_root *root, int fd)
     return heap_obj;
 
 fail_add_handle:
+	dma_buf_unmap_attachment_unlocked(heap_obj->import_attach, heap_obj->sgt, heap_obj->dir);
 fail_detach:
     dma_buf_detach(dma_buf, attach);
 clean_up:
@@ -289,7 +291,7 @@ static struct heap_mem *dmabuf_heap_import_with_dma_buf_st(struct heap_root *roo
         goto clean_up;
     }
 
-    sgt = dma_buf_map_attachment(attach, DMA_BIDIRECTIONAL);
+    sgt = dma_buf_map_attachment_unlocked(attach, DMA_BIDIRECTIONAL);
 	if (IS_ERR(sgt)) {
 		ret = PTR_ERR(sgt);
 		goto fail_detach;
@@ -321,6 +323,7 @@ static struct heap_mem *dmabuf_heap_import_with_dma_buf_st(struct heap_root *roo
     return heap_obj;
 
 fail_add_handle:
+	dma_buf_unmap_attachment_unlocked(heap_obj->import_attach, heap_obj->sgt, heap_obj->dir);
 fail_detach:
     dma_buf_detach(dma_buf, attach);
 clean_up:
@@ -398,7 +401,7 @@ static void __common_dmabuf_heap_release(struct kref *kref)
 
 	common_dmabuf_heap_umap_vaddr(heap_obj);
 
-	dma_buf_unmap_attachment(heap_obj->import_attach, heap_obj->sgt, heap_obj->dir);
+	dma_buf_unmap_attachment_unlocked(heap_obj->import_attach, heap_obj->sgt, heap_obj->dir);
 
 	dma_buf_detach(heap_obj->dbuf, heap_obj->import_attach);
 
@@ -430,7 +433,7 @@ void *common_dmabuf_heap_map_vaddr(struct heap_mem *heap_obj)
     if (heap_obj->vaddr)
         return heap_obj->vaddr;
 
-    ret = dma_buf_vmap(heap_obj->dbuf, &map);
+    ret = dma_buf_vmap_unlocked(heap_obj->dbuf, &map);
     if (ret)
         return NULL;
 
@@ -449,7 +452,7 @@ void common_dmabuf_heap_umap_vaddr(struct heap_mem *heap_obj)
 	if (heap_obj && heap_obj->vaddr) {
         map.vaddr = heap_obj->vaddr;
         map.is_iomem = 0;
-        dma_buf_vunmap(heap_obj->dbuf, &map);
+        dma_buf_vunmap_unlocked(heap_obj->dbuf, &map);
         heap_obj->vaddr = NULL;
     }
 }
diff --git a/drivers/soc/eswin/ai_driver/npu/npu_main.c b/drivers/soc/eswin/ai_driver/npu/npu_main.c
index e0099786ac87..6bb0a4a6a06c 100644
--- a/drivers/soc/eswin/ai_driver/npu/npu_main.c
+++ b/drivers/soc/eswin/ai_driver/npu/npu_main.c
@@ -73,7 +73,7 @@
 #include "mailbox_regs.h"
 #include "nvdla_proc.h"
 #include <linux/eswin_npu.h>
-
+#include <linux/dma-resv.h>
 #if defined(CONFIG_PM_DEVFREQ)
 #include <linux/devfreq.h>
 #include <linux/pm_opp.h>
@@ -123,7 +123,11 @@ int32_t dla_data_read(void *driver_context, void *task_data, void *handle,
 		dla_error("dma_buf_begin_cpu_access error\n");
 		goto put_dma_buf;
 	}
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 6, 0)
+	ret = dma_buf_vmap_unlocked(buf, &map);
+#else
 	ret = dma_buf_vmap(buf, &map);
+#endif
 	ptr = ret ? NULL : map.vaddr;
 	if (!ptr) {
 		pr_err("Failed to vmap dma_buf for fd=%d\n", fd);
@@ -139,7 +143,11 @@ int32_t dla_data_read(void *driver_context, void *task_data, void *handle,
 	} else {
 		memcpy(dst, (void *)(((uint8_t *)ptr) + offset), size);
 	}
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 6, 0)
+	dma_buf_vunmap_unlocked(buf, &map);
+#else
 	dma_buf_vunmap(buf, &map);
+#endif
 
 end_cpu_access:
 	dma_buf_end_cpu_access(buf, DMA_BIDIRECTIONAL);
@@ -349,7 +357,7 @@ static int npu_set_higher_vol_and_freq(struct nvdla_device *nvdla_dev, unsigned
 		dev_err(&nvdla_dev->pdev->dev, "get npu core clock pareent err.\n");
 		return -EINVAL;
 	}
-	
+
 	llc_rate = clk_get_rate(nvdla_dev->mux_u_npu_llclk_3mux1_gfree);
 	npu_rate = clk_get_rate(nvdla_dev->mux_u_npu_core_3mux1_gfree);
 
@@ -409,7 +417,7 @@ static int npu_set_lower_vol_and_freq(struct nvdla_device *nvdla_dev, unsigned l
 		dev_err(&nvdla_dev->pdev->dev, "get npu core clock pareent err.\n");
 		return -EINVAL;
 	}
-	
+
 	llc_rate = clk_get_rate(nvdla_dev->mux_u_npu_llclk_3mux1_gfree);
 	npu_rate = clk_get_rate(nvdla_dev->mux_u_npu_core_3mux1_gfree);
 
@@ -459,7 +467,7 @@ static int npu_devfreq_target(struct device *dev, unsigned long *freq, u32 flags
 	struct dev_pm_opp *opp;
 	unsigned long target_volt, target_rate;
 	int ret;
-	
+
 	if (nvdla_dev->is_low_freq) {
 		dev_err(dev, "Prohibit set voltage and freq, because dts set 'apply_npu_1G_freq' property.\n");
 		return 0;
@@ -495,7 +503,7 @@ static int npu_devfreq_target(struct device *dev, unsigned long *freq, u32 flags
 	        }
 
 	} else { // lower freq
-		
+
 		ret = npu_set_lower_vol_and_freq(nvdla_dev, target_rate);
 
 		if (ret) {
@@ -511,7 +519,7 @@ static int npu_devfreq_target(struct device *dev, unsigned long *freq, u32 flags
 		mdelay(10);
 
 	}
-        
+
 	nvdla_dev->rate = clk_get_rate(nvdla_dev->core_clk);
 	if (nvdla_dev->rate != target_rate) {
 		dev_err(dev, "Got wrong frequency, Request %lu, Current %lu.\n", target_rate, nvdla_dev->rate);
@@ -628,7 +636,7 @@ static int32_t edla_probe(struct platform_device *pdev)
 		nvdla_dev->npu_regulator = NULL;
 		return err;
 	}
-	
+
 	err = npu_dt_node_resources(nvdla_dev);
 	if (err) {
 		dla_error("error, get hw resource, ret=%d\n", err);
@@ -729,7 +737,7 @@ static int32_t edla_probe(struct platform_device *pdev)
 		dev_err(dev, "%s, %d, Failed to add OPP table, ret = %d.\n", __func__, __LINE__, err);
 		goto err_init_reset;
 	}
-	
+
 	df = devm_devfreq_add_device(dev, &npu_devfreq_profile, DEVFREQ_GOV_SIMPLE_ONDEMAND, &ondemand_data);
 	if (IS_ERR(df)) {
 		err = PTR_ERR(df);
diff --git a/drivers/staging/media/eswin/hae/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_dmabuf.c b/drivers/staging/media/eswin/hae/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_dmabuf.c
index 28b07d9b1431..4836c62ab6ac 100644
--- a/drivers/staging/media/eswin/hae/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_dmabuf.c
+++ b/drivers/staging/media/eswin/hae/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_dmabuf.c
@@ -217,7 +217,11 @@ _DmabufAttach(gckALLOCATOR Allocator, gcsATTACH_DESC_PTR Desc, PLINUX_MDL Mdl)
     if (!attachment)
         gcmkONERROR(gcvSTATUS_NOT_SUPPORTED);
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 6, 0)
+    sgt = dma_buf_map_attachment_unlocked(attachment, DMA_BIDIRECTIONAL);
+#else
     sgt = dma_buf_map_attachment(attachment, DMA_BIDIRECTIONAL);
+#endif
 
     if (!sgt)
         gcmkONERROR(gcvSTATUS_NOT_SUPPORTED);
@@ -288,7 +292,11 @@ _DmabufAttach(gckALLOCATOR Allocator, gcsATTACH_DESC_PTR Desc, PLINUX_MDL Mdl)
     }
 
     if (sgt)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 6, 0)
+        dma_buf_unmap_attachment_unlocked(attachment, sgt, DMA_BIDIRECTIONAL);
+#else
         dma_buf_unmap_attachment(attachment, sgt, DMA_BIDIRECTIONAL);
+#endif
 
     if (!os->device->platform->params.holdCrossAddr && dmabuf) {
         /** not release this dmabuf for keeping it as used,
@@ -312,7 +320,11 @@ _DmabufFree(gckALLOCATOR Allocator, PLINUX_MDL Mdl)
     list_del(&buf_desc->list);
     mutex_unlock(&priv->lock);
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 6, 0)
+    dma_buf_unmap_attachment_unlocked(buf_desc->attachment, buf_desc->sgt, DMA_BIDIRECTIONAL);
+#else
     dma_buf_unmap_attachment(buf_desc->attachment, buf_desc->sgt, DMA_BIDIRECTIONAL);
+#endif
 
     dma_buf_detach(buf_desc->dmabuf, buf_desc->attachment);
 
-- 
2.52.0

